storage/tests/test_isolation.rs

```diff
@@ -123,3 +127,488 @@ fn g0_update_conflicts_fail() {
+    let sum: u8 = reader_after.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_1)).unwrap().bytes()[0] + reader_after.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_2)).unwrap().bytes()[0];
+    assert!(sum <= 1);
+}
+
+#[test]
+fn g2_predicate_anti_dependency_cycles() {
+    // This also shows an rw-rw cycle, but with a predicate (realised as a iterate_prefix)
+    // t1 sums up a range, and adds an element to it. t2 sums up the same range and adds an element to it.
+    //      (Strictly, we only need to insert elements into the intersection of the ranges)
+    // One of their reads would not be repeatable under any serialisation.
+    init_logging();
+    let key_4_bytes = [0x0, 0x0, 0x0, 0x4];
+    let key_3 : StorageKeyArray<64>= StorageKeyArray::new(KEYSPACE_ID, ByteArray::copy(&KEY_3));
+    let key_4 : StorageKeyArray<64> = StorageKeyArray::new(KEYSPACE_ID, ByteArray::copy(&key_4_bytes));
+
+    let prefix: StorageKey<'_, 48> = StorageKey::Reference(StorageKeyReference::new(KEYSPACE_ID, ByteReference::new(&[0 as u8])));
+    let value_31 = ByteArray::inline([30], 1);
+    let value_42 = ByteArray::inline([42], 1);
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+
+
+    let it_1 = snapshot_1.iterate_prefix(prefix.clone());
+    let mut sum_1 : i64  = 0;
+    for z in it_1.collect_cloned_vec::<BUFFER_KEY_INLINE, BUFFER_VALUE_INLINE>().unwrap().iter() {
+        sum_1 += z.1.bytes()[0] as i64;
+    }
+    assert!(sum_1 == 1);
+
+    let it_2 = snapshot_2.iterate_prefix(prefix.clone());
+    let mut sum_2 : i64  = 0;
+    for z in it_2.collect_cloned_vec::<BUFFER_KEY_INLINE, BUFFER_VALUE_INLINE>().unwrap().iter() {
+        sum_2 += z.1.bytes()[0] as i64;
+    }
+    assert!(sum_2 == 1);
+
+    snapshot_1.put_val(key_3.to_owned(), ByteArray::copy(value_31.bytes()));
+    snapshot_2.put_val(key_4.to_owned(), ByteArray::copy(value_42.bytes()));
+    let result_1 = snapshot_1.commit();
+    let result_2 = snapshot_2.commit();
+
+    // These two rw anti-dependency edges are enough for the to not be serialisable.
+    // Both the assert sum_x == 1 cannot be true in any serialisation.
+    assert!(result_1.is_err() || result_2.is_err());
+}
+
+#[test]
+fn g2_antidependency_cycles_fekete() {
+    // From "A Read-Only Transaction Anomaly Under Snapshot Isolation" (Fekete et al.)
+    //  Without the read-only t3,  the history is: [ r_1(1), r_1(2), r_2(2), w_2(2, ?) ]
+    //      There is only one rw edge: r_1(2) -> w_2(2,?)
+    // The third transaction extends the history, [..., r_3(1), r_3(2), w_1(1, ?) ]
+    // opened after t2 commits introduces :
+    //  * a wr edge w_2(2,?) -> r_3(2)      (This is possible since t1 & t3 are not 'concurrent')
+    //  * a rw edge r_3(1) -> w_1(1,?)
+    // These form a cycle between t1, t2, t3.
```

Comment:
This is an example where the precedence graph cycle involves 3 transactions. This means we won't be able to detect this in a pairwise approach to conflict detection between transactions.